### PR TITLE
fix spawn MP context, limit cpu and download data

### PR DESCRIPTION
**Pull request by polisettyvarma**
https://github.com/HabanaAI/optimum-habana-fork/pull/274

### CODE REVIEW

File: examples/text-generation/requirements_lm_eval.txt [hunk 1]
```diff
@@ -1 +1 @@
-https://github.com/polisettyvarma/lm-evaluation-harness/archive/3cdc8daadad9f4559ae6cdfae96f1d83d6b3c1f4.zip
\ No newline at end of file
+git+https://github.com/EleutherAI/lm-evaluation-harness.git@master
```
Comment (yafshar) [inline examples/text-generation/requirements_lm_eval.txt:1]: @polisettyvarma please specify a compatible version (using semantic versioning) or use a branch-name or a working commit

File: examples/text-generation/run_lm_eval.py [hunk 2]
```diff
@@ -34,8 +34,29 @@
 
 
 os.environ.setdefault("TOKENIZERS_PARALLELISM", "false")
+os.environ.setdefault("HF_DATASETS_TRUST_REMOTE_CODE", "true")
```
Comment (yafshar) [inline examples/text-generation/run_lm_eval.py:37]: @polisettyvarma to be consistent with the rest of OH examples, why not add trust remote code as an input argument

File: examples/text-generation/requirements_lm_eval.txt (hunk 1 above)
Comment (polisettyvarma) [reply examples/text-generation/requirements_lm_eval.txt:1]: https://github.com/EleutherAI/lm-evaluation-harness/commits/master/ seems already frozen and no updates from past 7 to 8 months.

File: examples/text-generation/run_lm_eval.py (hunk 2 above)
Comment (polisettyvarma) [reply examples/text-generation/run_lm_eval.py]: @yafshar done, please take a look now

File: examples/text-generation/requirements_lm_eval.txt (hunk 1 above)
Comment (yafshar) [reply examples/text-generation/requirements_lm_eval.txt:1]: No matter that please follow the best practices!

File: examples/text-generation/run_lm_eval.py [hunk 3]
```diff
@@ -60,6 +80,8 @@ def setup_lm_eval_parser():
         default=["hellaswag", "lambada_openai", "piqa", "winogrande"],
     )
     parser.add_argument("--limit_iters", type=int, help="limit examples to run that many iterations", default=None)
+    parser.add_argument('--dont-trust-remote-datasets-code', action='store_false', default=True,
+                        dest='trust_remote_datasets_code', help='Trust remote code')
```
Comment (yafshar) [inline examples/text-generation/run_lm_eval.py:84]:
```suggestion
```

@polisettyvarma, you do not need this, you can use https://github.com/huggingface/optimum-habana/blob/c495f479d9abf04fb7adb6f0a5607d7963186649/examples/text-generation/run_generation.py#L285

File: examples/text-generation/run_lm_eval.py [hunk 4]
```diff
@@ -157,6 +179,8 @@ def _model_call(self, inps):
 
 def main():
     args = setup_lm_eval_parser()
+    if args.trust_remote_datasets_code:
```
Comment (yafshar) [inline examples/text-generation/run_lm_eval.py:182]:
```suggestion
    if args.trust_remote_code:
```

**Comment by yafshar** [comment]
@polisettyvarma, thanks for the modifications and this PR. To ensure we do not break anything, I will test the solution more thoroughly on other cases and update you!

**Comment by polisettyvarma** [comment]
> @polisettyvarma, thanks for the modifications and this PR. To ensure we do not break anything, I will test the solution more thoroughly on other cases and update you!

@yafshar i have update the patch, please check and use this for further testing

File: examples/text-generation/run_lm_eval.py [hunk 5]
```diff
@@ -157,6 +177,8 @@ def _model_call(self, inps):
 
 def main():
     args = setup_lm_eval_parser()
+    if args.trust_remote_code:
+        os.environ.setdefault("HF_DATASETS_TRUST_REMOTE_CODE", "true")
```
Comment (yafshar) [inline examples/text-generation/run_lm_eval.py:181]:
```suggestion
        os.environ.setdefault("HF_DATASETS_TRUST_REMOTE_CODE", "true")
        logger.info("Setting `HF_DATASETS_TRUST_REMOTE_CODE=true`")
```

**Comment by yafshar** [comment]
@polisettyvarma how did you test your implementation?

**Comment by polisettyvarma** [comment]
> @polisettyvarma how did you test your implementation? Can you provide an example?

@yafshar you mentioned about testing it, i thought you are aware of it. before and after my patch results should be similar.
https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation#lm-eval-requirements
https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation#examples
if you get issue about dataset download, need to pass "--trust-remote-code"

**Comment by yafshar** [comment]
> > @polisettyvarma how did you test your implementation? Can you provide an example?
> 
> @yafshar you mentioned about testing it, i thought you are aware of it. before and after my patch results should be similar. https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation#lm-eval-requirements https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation#examples if you get issue about dataset download, need to pass "--trust-remote-code"

@polisettyvarma thanks, I know how to test :) My question was that how did you evaluate or test your PR? I did not see any difference using your PR.  If I put a break-point in `LimitedSpawnPool` it never reaches that point! I have tried with and without deepspeed with 1 or multi tasks. Can you elaborate on your observations! Thanks

**Comment by polisettyvarma** [comment]
> > > @polisettyvarma how did you test your implementation? Can you provide an example?
> > 
> > 
> > @yafshar you mentioned about testing it, i thought you are aware of it. before and after my patch results should be similar. https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation#lm-eval-requirements https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation#examples if you get issue about dataset download, need to pass "--trust-remote-code"
> 
> @polisettyvarma thanks, I know how to test :) My question was that how did you evaluate or test your PR? I did not see any difference using your PR. If I put a break-point in `LimitedSpawnPool` it never reaches that point! I have tried with and without deepspeed with 1 or multi tasks. Can you elaborate on your observations! Thanks

@yafshar you can comment this line(https://github.com/huggingface/optimum-habana/pull/1131/files#diff-f1cc242d23dabd24c924359e72d7eba31a9edc5c6a589ca3ab8438d7fad8589bR57) and you should get segfault. keeping this line should resolve segfault

**Comment by yafshar** [comment]
> > > > @polisettyvarma how did you test your implementation? Can you provide an example?
> > > 
> > > 
> > > @yafshar you mentioned about testing it, i thought you are aware of it. before and after my patch results should be similar. https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation#lm-eval-requirements https://github.com/huggingface/optimum-habana/tree/main/examples/text-generation#examples if you get issue about dataset download, need to pass "--trust-remote-code"
> > 
> > 
> > @polisettyvarma thanks, I know how to test :) My question was that how did you evaluate or test your PR? I did not see any difference using your PR. If I put a break-point in `LimitedSpawnPool` it never reaches that point! I have tried with and without deepspeed with 1 or multi tasks. Can you elaborate on your observations! Thanks
> 
> @yafshar you can comment this line(https://github.com/huggingface/optimum-habana/pull/1131/files#diff-f1cc242d23dabd24c924359e72d7eba31a9edc5c6a589ca3ab8438d7fad8589bR57) and you should get segfault. keeping this line should resolve segfault

I do not get any segfault or any other issue. 

```bash
>> deepspeed --num_gpus 8 run_lm_eval.py --model_name_or_path meta-llama/Llama-2-70b-hf \
--use_hpu_graphs --use_kv_cache --bf16 --batch_size=1 --tasks winogrande -o eval.json

Running loglikelihood requests
100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 2534/2534 [04:35<00:00,  9.21it/s]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 2534/2534 [04:35<00:00,  9.21it/s]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 2534/2534 [04:35<00:00,  9.21it/s]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 2534/2534 [04:35<00:00,  9.21it/s]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 2534/2534 [04:35<00:00,  9.21it/s]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 2534/2534 [04:35<00:00,  9.21it/s]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 2534/2534 [04:35<00:00,  9.21it/s]
100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 2534/2534 [04:35<00:00,  9.21it/s]
Memory allocated                    = 18.6 GB
Max memory allocated                = 18.6 GB
Total memory available              = 94.62 GB
{
  "results": {
    "winogrande": {
      "acc": 0.8034727703235991,
      "acc_stderr": 0.011168120593569576
    }
  },
```

with and without comment, I do not see any differnce

```shell
OrigPool = mp.Pool
def LimitedSpawnPool(_):
    print(f"{'-'*100}\nInside LimitedSpawnPool\n{'-'*100}", flush=True)
    exit()
    spawn_context = mp.get_context("spawn")
    physical_cpu_count = psutil.cpu_count(logical=False)
    pool_size = physical_cpu_count
    world_size = int(os.getenv("WORLD_SIZE", 1))
    if world_size == 0:
        world_size = 1
    pool_size //= world_size
    if (pool_size * world_size) != physical_cpu_count:
        pool_size -= 1
    return spawn_context.Pool(pool_size)
mp.Pool = LimitedSpawnPool
```

**Comment by yafshar** [comment]
@polisettyvarma another point is that, for deepspeed use cases, it already has the options for `bind_cores_to_rank` & `bind_core_list` which is technically covering this PR intention. Unless you plan for other use cases which I am missing to see.

**Comment by polisettyvarma** [comment]
> @polisettyvarma another point is that, for deepspeed use cases, it already has the options for `bind_cores_to_rank` & `bind_core_list` which is technically covering this PR intention. Unless you plan for other use cases which I am missing to see.

@yafshar Thank you for testing. can you test for hellaswag task ?
another intention of this PR is to get rid of my fork and have transparent changes in code.

**Comment by yafshar** [comment]
@polisettyvarma thanks. I still do not see the issue!  I understand your intention, but do not see the point of it here. 

- If your point is using this option with DeepSpeed, then it already exists and the argument `--bind_cores_to_rank` in DeepSpeed, instructs DeepSpeed to split the CPU cores and distribute them to each rank evenly. This ensures that each inference worker uses an exclusive set of CPU cores to avoid interfering with one another.
- If you have other intentions, please state clearly and provide the use-cases, or examples

**Comment by yafshar** [comment]
@polisettyvarma if there is another point which is not covered by `--bind_cores_to_rank` in DeepSpeed, please state clearly and provide the use-cases, or examples, otherwise please close this PR. Thanks

**Comment by polisettyvarma** [comment]
> @polisettyvarma if there is another point which is not covered by `--bind_cores_to_rank` in DeepSpeed, please state clearly and provide the use-cases, or examples, otherwise please close this PR. Thanks

@yafshar please try to run this way and reproduce segfault
use this PR but comment the line58(https://github.com/huggingface/optimum-habana/pull/1131/files#diff-f1cc242d23dabd24c924359e72d7eba31a9edc5c6a589ca3ab8438d7fad8589bR58) to reproduce segfault and uncomment the same to resolve
```
cd <>/optimum-habana
pip install .
cd examples/text-generation
pip install -r requirements_lm_eval.txt
deepspeed --num_gpus 8 run_lm_eval.py --model_name_or_path "meta-llama/llama-2-7b-hf/" -o eval_output_70b_8x.txt --batch_size 8 --use_hpu_graphs --attn_softmax_bf16  --bucket_size=128 --bucket_internal  --task lambada_openai
```
please let me know the test results.

**Comment by yafshar** [comment]
> > @polisettyvarma if there is another point which is not covered by `--bind_cores_to_rank` in DeepSpeed, please state clearly and provide the use-cases, or examples, otherwise please close this PR. Thanks
> 
> @yafshar please try to run this way use this PR but comment the line58(https://github.com/huggingface/optimum-habana/pull/1131/files#diff-f1cc242d23dabd24c924359e72d7eba31a9edc5c6a589ca3ab8438d7fad8589bR58) to reproduce segfault and uncomment the same to resolve
> 
> ```
> cd <>/optimum-habana
> pip install .
> cd examples/text-generation
> pip install -r requirements_lm_eval.txt
> deepspeed --num_gpus 8 run_lm_eval.py --model_name_or_path "meta-llama/llama-2-7b-hf/" -o eval_output_70b_8x.txt --batch_size 8 --use_hpu_graphs --attn_softmax_bf16  --bucket_size=128 --bucket_internal  --task lambada_openai
> ```
> 
> please let me know the test results.

@polisettyvarma I see no difference and am not able to reproduce the issue!

I did three tests on 1.16.0 & 1.17.0
 
Driver Version: 1.16.0-94aac46

with your PR
```sh
Memory allocated                    = 6.06 GB
Max memory allocated                = 6.06 GB
Total memory available              = 94.62 GB
{
  "results": {
    "lambada_openai": {
      "ppl": 3.4206070920962195,
      "ppl_stderr": 0.06678639625173022,
      "acc": 0.7323888996700951,
      "acc_stderr": 0.006167867157277631
    }
  },
  "versions": {
    "lambada_openai": 0
  },
  "args": {
    "buckets": [
      16,
      32,
      64,
      128,
      189,
      284
    ],
    "output_file": "eval_output_70b_8x.txt",
    "tasks": [
      "lambada_openai"
    ],
    "limit_iters": null,
    "device": "hpu",
    "model_name_or_path": "meta-llama/Llama-2-7b-hf",
    "bf16": false,
    "max_new_tokens": 100,
    "max_input_tokens": 0,
    "batch_size": 8,
    "warmup": 3,
    "n_iterations": 5,
    "local_rank": 0,
    "use_kv_cache": false,
    "use_hpu_graphs": true,
    "dataset_name": null,
    "column_name": null,
    "do_sample": false,
    "num_beams": 1,
    "trim_logits": false,
    "seed": 27,
    "profiling_warmup_steps": 0,
    "profiling_steps": 0,
    "profiling_record_shapes": false,
    "prompt": null,
    "bad_words": null,
    "force_words": null,
    "assistant_model": null,
    "peft_model": null,
    "num_return_sequences": 1,
    "token": null,
    "model_revision": "main",
    "attn_softmax_bf16": true,
    "output_dir": null,
    "bucket_size": 128,
    "bucket_internal": true,
    "dataset_max_samples": -1,
    "limit_hpu_graphs": false,
    "reuse_cache": false,
    "verbose_workers": false,
    "simulate_dyn_prompt": null,
    "reduce_recompile": false,
    "use_flash_attention": false,
    "flash_attention_recompute": false,
    "flash_attention_causal_mask": false,
    "flash_attention_fast_softmax": false,
    "book_source": false,
    "torch_compile": false,
    "ignore_eos": true,
    "temperature": 1.0,
    "top_p": 1.0,
    "const_serialization_path": null,
    "disk_offload": false,
    "trust_remote_code": false,
    "quant_config": "",
    "world_size": 8,
    "global_rank": 0
  },
  "duration": 238.44612671100185
}
[2024-07-26 16:47:17,016] [INFO] [launch.py:348:main] Process 32141 exits successfully.
[2024-07-26 16:47:18,017] [INFO] [launch.py:348:main] Process 32144 exits successfully.
[2024-07-26 16:47:18,018] [INFO] [launch.py:348:main] Process 32140 exits successfully.
[2024-07-26 16:47:18,018] [INFO] [launch.py:348:main] Process 32147 exits successfully.
[2024-07-26 16:47:18,018] [INFO] [launch.py:348:main] Process 32146 exits successfully.
[2024-07-26 16:47:18,018] [INFO] [launch.py:348:main] Process 32145 exits successfully.
[2024-07-26 16:47:18,018] [INFO] [launch.py:348:main] Process 32143 exits successfully.
[2024-07-26 16:47:18,018] [INFO] [launch.py:348:main] Process 32142 exits successfully.
```

with commenting line 58
```sh
Memory allocated                    = 6.06 GB
Max memory allocated                = 6.06 GB
Total memory available              = 94.62 GB
{
  "results": {
    "lambada_openai": {
      "ppl": 3.4206070920962195,
      "ppl_stderr": 0.06678639625173022,
      "acc": 0.7323888996700951,
      "acc_stderr": 0.006167867157277631
    }
  },
  "versions": {
    "lambada_openai": 0
  },
  "args": {
    "buckets": [
      16,
      32,
      64,
      128,
      189,
      284
    ],
    "output_file": "eval_output_70b_8x.txt",
    "tasks": [
      "lambada_openai"
    ],
    "limit_iters": null,
    "device": "hpu",
    "model_name_or_path": "meta-llama/Llama-2-7b-hf",
    "bf16": false,
    "max_new_tokens": 100,
    "max_input_tokens": 0,
    "batch_size": 8,
    "warmup": 3,
    "n_iterations": 5,
    "local_rank": 0,
    "use_kv_cache": false,
    "use_hpu_graphs": true,
    "dataset_name": null,
    "column_name": null,
    "do_sample": false,
    "num_beams": 1,
    "trim_logits": false,
    "seed": 27,
    "profiling_warmup_steps": 0,
    "profiling_steps": 0,
    "profiling_record_shapes": false,
    "prompt": null,
    "bad_words": null,
    "force_words": null,
    "assistant_model": null,
    "peft_model": null,
    "num_return_sequences": 1,
    "token": null,
    "model_revision": "main",
    "attn_softmax_bf16": true,
    "output_dir": null,
    "bucket_size": 128,
    "bucket_internal": true,
    "dataset_max_samples": -1,
    "limit_hpu_graphs": false,
    "reuse_cache": false,
    "verbose_workers": false,
    "simulate_dyn_prompt": null,
    "reduce_recompile": false,
    "use_flash_attention": false,
    "flash_attention_recompute": false,
    "flash_attention_causal_mask": false,
    "flash_attention_fast_softmax": false,
    "book_source": false,
    "torch_compile": false,
    "ignore_eos": true,
    "temperature": 1.0,
    "top_p": 1.0,
    "const_serialization_path": null,
    "disk_offload": false,
    "trust_remote_code": false,
    "quant_config": "",
    "world_size": 8,
    "global_rank": 0
  },
  "duration": 240.20362453197595
}
[2024-07-26 16:53:49,176] [INFO] [launch.py:348:main] Process 55712 exits successfully.
[2024-07-26 16:53:49,177] [INFO] [launch.py:348:main] Process 55708 exits successfully.
[2024-07-26 16:53:49,177] [INFO] [launch.py:348:main] Process 55715 exits successfully.
[2024-07-26 16:53:49,177] [INFO] [launch.py:348:main] Process 55714 exits successfully.
[2024-07-26 16:53:49,177] [INFO] [launch.py:348:main] Process 55713 exits successfully.
[2024-07-26 16:53:49,177] [INFO] [launch.py:348:main] Process 55711 exits successfully.
[2024-07-26 16:53:49,177] [INFO] [launch.py:348:main] Process 55709 exits successfully.
[2024-07-26 16:53:49,177] [INFO] [launch.py:348:main] Process 55710 exits successfully.
```

with optimum-habana main branch
```sh
Memory allocated                    = 6.06 GB
Max memory allocated                = 6.06 GB
Total memory available              = 94.62 GB
{
  "results": {
    "lambada_openai": {
      "ppl": 3.4206070920962195,
      "ppl_stderr": 0.06678639625173022,
      "acc": 0.7323888996700951,
      "acc_stderr": 0.006167867157277631
    }
  },
  "versions": {
    "lambada_openai": 0
  },
  "args": {
    "buckets": [
      16,
      32,
      64,
      128,
      189,
      284
    ],
    "output_file": "eval_output_70b_8x.txt",
    "tasks": [
      "lambada_openai"
    ],
    "limit_iters": null,
    "device": "hpu",
    "model_name_or_path": "meta-llama/Llama-2-7b-hf",
    "bf16": false,
    "max_new_tokens": 100,
    "max_input_tokens": 0,
    "batch_size": 8,
    "warmup": 3,
    "n_iterations": 5,
    "local_rank": 0,
    "use_kv_cache": false,
    "use_hpu_graphs": true,
    "dataset_name": null,
    "column_name": null,
    "do_sample": false,
    "num_beams": 1,
    "trim_logits": false,
    "seed": 27,
    "profiling_warmup_steps": 0,
    "profiling_steps": 0,
    "profiling_record_shapes": false,
    "prompt": null,
    "bad_words": null,
    "force_words": null,
    "assistant_model": null,
    "peft_model": null,
    "num_return_sequences": 1,
    "token": null,
    "model_revision": "main",
    "attn_softmax_bf16": true,
    "output_dir": null,
    "bucket_size": 128,
    "bucket_internal": true,
    "dataset_max_samples": -1,
    "limit_hpu_graphs": false,
    "reuse_cache": false,
    "verbose_workers": false,
    "simulate_dyn_prompt": null,
    "reduce_recompile": false,
    "use_flash_attention": false,
    "flash_attention_recompute": false,
    "flash_attention_causal_mask": false,
    "flash_attention_fast_softmax": false,
    "book_source": false,
    "torch_compile": false,
    "ignore_eos": true,
    "temperature": 1.0,
    "top_p": 1.0,
    "const_serialization_path": null,
    "disk_offload": false,
    "trust_remote_code": false,
    "quant_config": "",
    "world_size": 8,
    "global_rank": 0
  },
  "duration": 241.29430994499126
}
[2024-07-26 17:03:35,551] [INFO] [launch.py:348:main] Process 79652 exits successfully.
[2024-07-26 17:03:35,552] [INFO] [launch.py:348:main] Process 79659 exits successfully.
[2024-07-26 17:03:35,552] [INFO] [launch.py:348:main] Process 79658 exits successfully.
[2024-07-26 17:03:35,552] [INFO] [launch.py:348:main] Process 79657 exits successfully.
[2024-07-26 17:03:35,552] [INFO] [launch.py:348:main] Process 79655 exits successfully.
[2024-07-26 17:03:35,552] [INFO] [launch.py:348:main] Process 79653 exits successfully.
[2024-07-26 17:03:35,552] [INFO] [launch.py:348:main] Process 79654 exits successfully.
[2024-07-26 17:03:35,552] [INFO] [launch.py:348:main] Process 79656 exits successfully.
```

**Comment by yafshar** [comment]
@polisettyvarma below is the result of `deepspeed --bind_cores_to_rank`. This approach is the standard way to go also see the duration `"duration": 88.44215309299761` vs your PR `"duration": 238.44612671100185`

```sh
deepspeed --bind_cores_to_rank --num_gpus 8 run_lm_eval.py --model_name_or_path meta-llama/Llama-2-7b-hf -o eval_output_70b_8x.txt --batch_size 8 --use_hpu_graphs --attn_softmax_bf16  --bucket_size=128 --bucket_internal  --task lambada_openai
```

```sh
Memory allocated                    = 6.06 GB
Max memory allocated                = 6.06 GB
Total memory available              = 94.62 GB
{
  "results": {
    "lambada_openai": {
      "ppl": 3.4206070920962195,
      "ppl_stderr": 0.06678639625173022,
      "acc": 0.7323888996700951,
      "acc_stderr": 0.006167867157277631
    }
  },
  "versions": {
    "lambada_openai": 0
  },
  "args": {
    "buckets": [
      16,
      32,
      64,
      128,
      189,
      284
    ],
    "output_file": "eval_output_70b_8x.txt",
    "tasks": [
      "lambada_openai"
    ],
    "limit_iters": null,
    "device": "hpu",
    "model_name_or_path": "meta-llama/Llama-2-7b-hf",
    "bf16": false,
    "max_new_tokens": 100,
    "max_input_tokens": 0,
    "batch_size": 8,
    "warmup": 3,
    "n_iterations": 5,
    "local_rank": 0,
    "use_kv_cache": false,
    "use_hpu_graphs": true,
    "dataset_name": null,
    "column_name": null,
    "do_sample": false,
    "num_beams": 1,
    "trim_logits": false,
    "seed": 27,
    "profiling_warmup_steps": 0,
    "profiling_steps": 0,
    "profiling_record_shapes": false,
    "prompt": null,
    "bad_words": null,
    "force_words": null,
    "assistant_model": null,
    "peft_model": null,
    "num_return_sequences": 1,
    "token": null,
    "model_revision": "main",
    "attn_softmax_bf16": true,
    "output_dir": null,
    "bucket_size": 128,
    "bucket_internal": true,
    "dataset_max_samples": -1,
    "limit_hpu_graphs": false,
    "reuse_cache": false,
    "verbose_workers": false,
    "simulate_dyn_prompt": null,
    "reduce_recompile": false,
    "use_flash_attention": false,
    "flash_attention_recompute": false,
    "flash_attention_causal_mask": false,
    "flash_attention_fast_softmax": false,
    "book_source": false,
    "torch_compile": false,
    "ignore_eos": true,
    "temperature": 1.0,
    "top_p": 1.0,
    "const_serialization_path": null,
    "disk_offload": false,
    "trust_remote_code": false,
    "quant_config": "",
    "world_size": 8,
    "global_rank": 0
  },
  "duration": 88.44215309299761
}
[2024-07-26 17:31:57,478] [INFO] [launch.py:348:main] Process 104902 exits successfully.
[2024-07-26 17:31:57,478] [INFO] [launch.py:348:main] Process 104915 exits successfully.
[2024-07-26 17:31:57,478] [INFO] [launch.py:348:main] Process 104898 exits successfully.
[2024-07-26 17:31:57,478] [INFO] [launch.py:348:main] Process 104967 exits successfully.
[2024-07-26 17:31:57,478] [INFO] [launch.py:348:main] Process 104941 exits successfully.
[2024-07-26 17:31:57,478] [INFO] [launch.py:348:main] Process 104928 exits successfully.
[2024-07-26 17:31:57,478] [INFO] [launch.py:348:main] Process 104947 exits successfully.
[2024-07-26 17:31:57,479] [INFO] [launch.py:348:main] Process 104980 exits successfully.
```

**Comment by yafshar** [comment]
@polisettyvarma  what you are doing in this PR `pool_size //= world_size; spawn_context.Pool(pool_size)` is done in a correct way using `numactl` and with core affinity in DeepSpeed `deepspeed --bind_cores_to_rank` . 
I do not see any other point for your PR and you are not able to demonstrate any other reason for this PR. DeepSpeed is doing exactly what you want to do.
Please close this PR.

**Comment by polisettyvarma** [comment]
> @polisettyvarma what you are doing in this PR `pool_size //= world_size; spawn_context.Pool(pool_size)` is done in a correct way using `numactl` and with core affinity in DeepSpeed `deepspeed --bind_cores_to_rank` . I do not see any other point for your PR and you are not able to demonstrate any other reason for this PR. DeepSpeed is doing exactly what you want to do. Please close this PR.

This PR doesn't intend to improve any performance but to get rid of my personal lm-eval fork, instead use master with specific commit and monkey patching to resolve segfault

**Comment by yafshar** [comment]
> > @polisettyvarma what you are doing in this PR `pool_size //= world_size; spawn_context.Pool(pool_size)` is done in a correct way using `numactl` and with core affinity in DeepSpeed `deepspeed --bind_cores_to_rank` . I do not see any other point for your PR and you are not able to demonstrate any other reason for this PR. DeepSpeed is doing exactly what you want to do. Please close this PR.
> 
> This PR doesn't intend to improve any performance but to get rid of my personal lm-eval fork, instead use master with specific commit and monkey patching to resolve segfault

Can you please confirm that if you do `deepspeed --bind_cores_to_rank` it does not resolve your issue! and please post the error that you are getting which necessitate to have this patch in the general repository that not only you but many others will use.

**Comment by yafshar** [comment]
@libinta @regisss I do not see any point of having this implementation in the code. From my experience and as I explained above `deepspeed --bind_cores_to_rank` does exactly as author intended with advantage of core affinity + more finer control to the users with extra option.

**Review by yafshar — APPROVED** [review]
I reviewed this PR with @polisettyvarma, and was able to reproduce the issue. It might happen on special cases, also if the user uses torch run. Thus it makes sense to have the change here.

LGTM!

@regisss please look at this PR.

**Pull request merged**